File: 0001-0100/0076-Minimum Window Substring/0076-Minimum Window Substring.py
import collections
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        table = collections.Counter(t)
        count = len(table)
        start = 0
        minSize = len(s) + 1
        result = ''
        for i in range(len(s)):
            table[s[i]] -= 1
            if table[s[i]] == 0:
                count -= 1
                while count == 0:
                    if i - start + 1 < minSize:
                        minSize = i - start + 1
                        result = s[start: i+1]
                    table[s[start]] += 1
                    if table[s[start]] == 1:
                        count += 1
                    start += 1

        return result

File: 0001-0100/0076-Minimum Window Substring/0076-Minimum Window Substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char, int> table;
        for (char c : t) {
            ++table[c];
        }
        int count = table.size(), start = 0;
        int minSize = s.size() + 1;
        string result;
        for (int i = 0; i < s.size(); ++i) {
            if (--table[s[i]] == 0) {
                --count;
            }
            while (count == 0) {
                if (i - start + 1 < minSize) {
                    minSize = i - start + 1;
                    result = s.substr(start, minSize);
                }
                if (++table[s[start++]] == 1) {
                    ++count;
                }
            }
        }
        
        return result;
    }
};
